test/test_autocast_xla.py

```diff
@@ -0,0 +1,37 @@
+import re
+import torch
+import torch_xla
+import torch_xla.core.xla_model as xm
+import unittest
+
+device = xm.xla_device()
+
+
+class TestAutocastXla(unittest.TestCase):
+
+  def test_cross_entropy_loss(self):
+    data = torch.randn(16, 10).to(torch.bfloat16).to(device)
+    target = torch.randn(16, 10).to(torch.bfloat16).to(device)
+
+    with torch.autocast("xla"):
+      loss = torch.nn.CrossEntropyLoss()(data, target)
+      hlo = torch_xla._XLAC._get_xla_tensors_hlo([loss])
+      self.assertRegex(hlo, r".*convert.*f32.*convert.*bf16")
+      self.assertRegex(hlo, r".*exponential.*f32.*exponential.*f32")
+      self.assertRegex(hlo, r".*log.*f32.*log.*f32")
+
+  def test_einsum(self):
+    data = torch.randn(16, 10).to(torch.bfloat16).to(device)
```

Comment:
Just curious what would happen if the input data was fp32 in this case. Do we still expect the output to be bf16? If yes, can we add a test case for that?

Reply:
@bhavya01 I verified what @avizon-aws stated above, and indeed irrespective of input data being bf16 or fp32, autocast would get output in bf16 (by casting the inputs to bf16 if necessary), as set in the autocast policy.

Adding this as a testcase, and attaching HLO's generated below for reference.

HLO with bf16 outputs:
```
HloModule IrToHlo.6, entry_computation_layout={(bf16[5,10]{1,0}, bf16[16,10]{1,0})->(bf16[16,5]{1,0})}

ENTRY %IrToHlo.6 (p0.1: bf16[5,10], p1.2: bf16[16,10]) -> (bf16[16,5]) {
  %p1.2 = bf16[16,10]{1,0} parameter(1), metadata={op_type="xla__device_data" op_name="xla__device_data" source_file="/shared_new/nmokashi/workspace/opensource_repos/pytorch_xla/xla/test/test_bf16_autocast.py" source_line=26}
  %p0.1 = bf16[5,10]{1,0} parameter(0), metadata={op_type="xla__device_data" op_name="xla__device_data" source_file="/shared_new/nmokashi/workspace/opensource_repos/pytorch_xla/xla/test/test_bf16_autocast.py" source_line=27}
  %dot.3 = bf16[16,5]{1,0} dot(bf16[16,10]{1,0} %p1.2, bf16[5,10]{1,0} %p0.1), lhs_contracting_dims={1}, rhs_contracting_dims={1}, frontend_attributes={grad_x="false",grad_y="false"}, metadata={op_type="aten__einsum" op_name="aten__einsum" source_file="/shared_new/nmokashi/testenv_ptGSPMD_6271032615/lib/python3.10/site-packages/torch/functional.py" source_line=407}
  %transpose.4 = bf16[16,5]{1,0} transpose(bf16[16,5]{1,0} %dot.3), dimensions={0,1}, metadata={op_type="aten__einsum" op_name="aten__einsum" source_file="/shared_new/nmokashi/testenv_ptGSPMD_6271032615/lib/python3.10/site-packages/torch/functional.py" source_line=407}
  ROOT %tuple.5 = (bf16[16,5]{1,0}) tuple(bf16[16,5]{1,0} %transpose.4)
}
```

HLO with f32 outputs:
```
HloModule IrToHlo.8, entry_computation_layout={(f32[5,10]{1,0}, f32[16,10]{1,0})->(bf16[16,5]{1,0})}

ENTRY %IrToHlo.8 (p0.1: f32[5,10], p1.3: f32[16,10]) -> (bf16[16,5]) {
  %p1.3 = f32[16,10]{1,0} parameter(1), metadata={op_type="xla__device_data" op_name="xla__device_data" source_file="/shared_new/nmokashi/workspace/opensource_repos/pytorch_xla/xla/test/test_bf16_autocast.py" source_line=29}
  %convert.4 = bf16[16,10]{1,0} convert(f32[16,10]{1,0} %p1.3), metadata={op_type="xla__cast" op_name="xla__cast" source_file="/shared_new/nmokashi/testenv_ptGSPMD_6271032615/lib/python3.10/site-packages/torch/functional.py" source_line=407}
  %p0.1 = f32[5,10]{1,0} parameter(0), metadata={op_type="xla__device_data" op_name="xla__device_data" source_file="/shared_new/nmokashi/workspace/opensource_repos/pytorch_xla/xla/test/test_bf16_autocast.py" source_line=30}
  %convert.2 = bf16[5,10]{1,0} convert(f32[5,10]{1,0} %p0.1), metadata={op_type="xla__cast" op_name="xla__cast" source_file="/shared_new/nmokashi/testenv_ptGSPMD_6271032615/lib/python3.10/site-packages/torch/functional.py" source_line=407}
  %dot.5 = bf16[16,5]{1,0} dot(bf16[16,10]{1,0} %convert.4, bf16[5,10]{1,0} %convert.2), lhs_contracting_dims={1}, rhs_contracting_dims={1}, frontend_attributes={grad_x="false",grad_y="false"}, metadata={op_type="aten__einsum" op_name="aten__einsum" source_file="/shared_new/nmokashi/testenv_ptGSPMD_6271032615/lib/python3.10/site-packages/torch/functional.py" source_line=407}
  %transpose.6 = bf16[16,5]{1,0} transpose(bf16[16,5]{1,0} %dot.5), dimensions={0,1}, metadata={op_type="aten__einsum" op_name="aten__einsum" source_file="/shared_new/nmokashi/testenv_ptGSPMD_6271032615/lib/python3.10/site-packages/torch/functional.py" source_line=407}
  ROOT %tuple.7 = (bf16[16,5]{1,0}) tuple(bf16[16,5]{1,0} %transpose.6)
}
```